process/quic.cpp

```diff
@@ -0,0 +1,876 @@
+   EVP_CIPHER_CTX *ctx;
+
+   if (!(ctx = EVP_CIPHER_CTX_new())) {
+      DEBUG_MSG("Sample encryption, creating context failed\n");
+      return false;
+   }
+   if (!(EVP_EncryptInit_ex(ctx, EVP_aes_128_ecb(), NULL, initial_secrets.hp, NULL))) {
+      DEBUG_MSG("Sample encryption, context initialization failed\n");
+      EVP_CIPHER_CTX_free(ctx);
+      return false;
+   }
+
+
+   // set padding always returns 1 so no need for success
+   // we need to disable padding so we can use EncryptFinal
+
+   EVP_CIPHER_CTX_set_padding(ctx, 0);
+   if (!(EVP_EncryptUpdate(ctx, plaintext, &len, sample, SAMPLE_LENGTH))) {
+      DEBUG_MSG("Sample encryption, decrypting header failed\n");
+      EVP_CIPHER_CTX_free(ctx);
+      return false;
+   }
+   if (!(EVP_EncryptFinal_ex(ctx, plaintext + len, &len))) {
+      DEBUG_MSG("Sample encryption, final header decryption failed\n");
+      EVP_CIPHER_CTX_free(ctx);
+      return false;
+   }
+
+   EVP_CIPHER_CTX_free(ctx);
+   memcpy(mask, plaintext, sizeof(mask));
+
+   // https://datatracker.ietf.org/doc/html/draft-ietf-quic-tls-22#section-5.4.1
+
+   //   if (packet[0] & 0x80) == 0x80:
+   //      # Long header: 4 bits masked
+   //      packet[0] ^= mask[0] & 0x0f
+   //   else:
+   //     # Short header: 5 bits masked
+   //     packet[0] ^= mask[0] & 0x1f
+
+   // we do not have to handle short header, Initial packets have only long header
+
+   first_byte  = quic_h1->first_byte;
+   first_byte ^= mask[0] & 0x0f;
+   uint8_t pkn_len = (first_byte & 0x03) + 1;
+
+   // set decrypted first byte
+   header[0] = first_byte;
+
+
+   // copy encrypted pkn into buffer
+   memcpy(&full_pkn, pkn, pkn_len);
+
+
+   // decrypt pkn
+   for (unsigned int i = 0; i < pkn_len; i++) {
+      packet_number |= (full_pkn[i] ^ mask[1 + i]) << (8 * (pkn_len - 1 - i));
+   }
+
+
+   // after decrypting first byte, we know packet number length, so we can adjust payload start and lengths
+   payload     = payload + pkn_len;
+   payload_len = payload_len - pkn_len;
+
+   // SET HEADER LENGTH, if header length is set incorrectly AEAD will calculate wrong tag, so decryption will fail
+   header_len = payload - header;
+
+   // set decrypted packet number
+   for (unsigned i = 0; i < pkn_len; i++) {
+      header[header_len - 1 - i] = (uint8_t) (packet_number >> (8 * i));
+   }
+
+
+   // adjust nonce for payload decryption
+   phton64(nonce + sizeof(nonce) - 8, pntoh64(nonce + sizeof(nonce) - 8) ^ packet_number);
+
+   return true;
+} // QUICPlugin::quic_decrypt_header
+
+void QUICPlugin::phton64(uint8_t *p, uint64_t v)
+{
+   int shift = 56;
+
+   for (unsigned int i = 0; i < 8; i++) {
+      p[i] = (uint8_t) (v >> (shift - (i * 8)));
+   }
+}
+
+uint64_t QUICPlugin::pntoh64(const void *p)
+{
+   uint64_t buffer = 0;
+   int shift       = 56;
+
+   for (unsigned x = 0; x < 8; x++) {
+      buffer |= (uint64_t) *((const uint8_t *) (p) + x) << (shift - (x * 8));
+   }
+   return buffer;
+}
```

Comment:
Shouldn'ŧ we put these functions somewhere into utils / common / ? I.e., somewhere outside the plugin?